hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/CryptoTransferHandler.java

```diff
@@ -616,10 +629,100 @@ public Fees calculateFees(@NonNull final FeeContext feeContext) {
         return feeContext
                 .feeCalculator(subType)
                 .addBytesPerTransaction(bpt)
+                .addRamByteSeconds(associationsRbs)
                 .addRamByteSeconds(rbs * USAGE_PROPERTIES.legacyReceiptStorageSecs())
                 .calculate();
     }
 
+    private long getAutoAssociationsRbsFee(final FeeContext feeContext) {
+        final var body = feeContext.body();
+        final var op = body.cryptoTransferOrThrow();
+        final var tokenTransfers = op.tokenTransfers();
+
+        // If the token transfers collection is empty, then we have only hbars transfer and no token
+        // associations are needed, so we do not need to calculate the next part.
+        if (tokenTransfers.isEmpty()) {
+            return 0;
+        }
+
+        final var accountStore = feeContext.readableStore(ReadableAccountStore.class);
+        final var tokenRelStore = feeContext.readableStore(ReadableTokenRelationStore.class);
+
+        final Set<AccountID> recipientsSet = getTransferRecipients(tokenTransfers);
+        final HashMap<AccountID, HashSet<TokenID>> associationsMap = getTokenTransferAssociations(tokenTransfers);
+
+        var associationsRbs = 0;
+        for (var accountId : recipientsSet) {
+            var account = accountStore.getAccountById(accountId);
+            // If this is a transfer to an alias and the account is null this is a hollow account creation case,
+            // and we need to set default values.
+            if (account == null && accountId.hasAlias()) {
+                var numAutoAssociations = associationsMap.get(accountId).size();
+                associationsRbs += numAutoAssociations * THREE_MONTHS_IN_SECONDS * CREATE_SLOT_MULTIPLIER;
+            } else {
+                var tokenIds = associationsMap.get(accountId);
+                var newAutoAssociations = 0;
+                for (var tokenId : tokenIds) {
+                    var tokenRel = tokenRelStore.get(accountId, tokenId);
+                    // If there already is an association, no need to compute further association fees.
+                    if (tokenRel != null) {
+                        continue;
+                    }
+                    // If there is no association, we need to compute the fee to create it.
+                    // Case 1: maxAutoAssociations == -1 -> unlimited auto associations, just increase newAutoAssociations counter
+                    // Case 2: maxAutoAssociations == 0 -> no auto associations can be made
+                    // Case 3: maxAutoAssociations > 0 -> we need to check if there are unused slots
+                    if (account.maxAutoAssociations() >= 0) {
+                        validateFalse(
+                                account.usedAutoAssociations() + 1 > account.maxAutoAssociations(),
+                                NO_REMAINING_AUTOMATIC_ASSOCIATIONS);
+                    }
```

Comment:
Why should we check this ? If unlimitedAutoAssociationsIdEnabled, we should just charge for new associations right ?